# Preprocessing Nextflow workflow

This scRNA-Seq preocessing Nextflow workflow is part of the Single cell RNASeq workflow developed for the Warren Lab. The following document outlines the input and output specifications, data formatting requirements and runtime enviroment details for the workflow. The preprocesing nextflow performs data QC, filters low quality cells, incorporate VDJ data when available, generate SCE objects compatible with scanpy, Seurat, and Monocle3. 

## Inputs:

All the inputs for the analysis are read out of a config file, the structure of the config file is listed below

```{nextflow}
params {

    input {
      count_dir = "/home/sravisha/fuser/nextflow/ZhengSorted_10X/Counts/"
      vdj_dir = false
      sample_list = ['S1_Bcells', 'S1_CytotoxicTCells', 'S1_HelperTCells', 'S1_MemoryTCells', 'S1_Monocytes', 'S1_NaiveCytotoxicTCells', 'S1_NaiveTCells', 'S1_NKCells', 'S1_Progenitor', 'S1_RegulatoryTCells']
    }
    preprocess {
      nmads = 3 
    }
    processmeta {
      vdj = false   
      gex_metadata = "/home/sravisha/fuser/nextflow/ZhengSorted_10X/Metadata/GEX_h5_samplesheet.csv"
      vdj_metadata = false
    }
    output {
      folder = "/home/sravisha/fuser/nextflow/ZhengSorted_10X/"
    }

}
```

The parameters file is divided into four sections.

1. Input

   The input section contains the path to the count directory generated by the cellranger nextflow, and a list of sample names that should be included for processing and downstream analysis.

   i. Count directory

       Path to count matrices generated using the cellranger nextflow workflow. 

   ii. VDJ directory

       Path to results from VDJ workflow from the cellranger nextflow. 
   
   iii. Sample list
       
       List of sample names that should be included for the downstream analysis.
      
2. Output
   
   All results from the cellranger Nextflow will be stored within subfolders under this output path.

3. Preprocess

   The preprocessing parameter required for the workflow is the number of median absolute deviations away from median distribution gene expression to identify outliers. 
4. Aggregate

   Specifies what type of normalization should be performed during the aggregation step.
    

## Outputs:

The pipeline generates three folders in the output folder:

1. Counts

   Contains the count matrices for each samples in the analysis. This serves as the input for most of the other nextflow scripts for downstream analysis

2. Metadata
   
   The metadata folder contain the samplesheets for the VDJ and GEX analysisas well as additional information such files paths for count matrices and VDJ clonotypes and contig files. These sample metadta files serve as the input to the preprocessing nextflow that generate the Single Cell Experiment objects for each sample. All information from the sample metadata files is added to the colData of the Single Cell Experiment√ü

3. VDJ
   
   The VDJ folder contains the results from the cellranger VDJ pipeline when available. The folder contains the clonotypes and contig. When avaiable the VDJ data is incorporated into the Single Cell Experiment object in the preprocessing nextflow

